stdlib/seq.mli

```diff
@@ -13,97 +13,807 @@
+
+val repeat : 'a -> 'a t
+(** [repeat x] is the infinite sequence
+    where the element [x] is repeated indefinitely.
+
+    [repeat x] is equivalent to [cycle (return x)].
+
+    @since 4.14 *)
+
+val forever : (unit -> 'a) -> 'a t
+(** [forever f] is an infinite sequence where every element is produced
+    (on demand) by the function call [f()].
+
+    For instance,
+    [forever Random.bool] is an infinite sequence of random bits.
+
+    [forever f] is equivalent to [map f (repeat ())].
+
+    @since 4.14 *)
+
+val cycle : 'a t -> 'a t
+(** [cycle xs] is the infinite sequence which consists of an infinite
+    number of repetitions of the sequence [xs].
+
+    The sequence [xs] must be nonempty.
+
+    Consuming (a prefix of) the sequence [cycle xs] once
+    can cause the sequence [xs] to be consumed more than once.
+    Therefore, [xs] must be persistent.
+
+    @raise Invalid_argument if [xs] is empty.
+    @since 4.14 *)
+
+val iterate : ('a -> 'a) -> 'a -> 'a t
+(** [iterate f x] is the infinite sequence whose elements are
+    [x], [f x], [f (f x)], and so on.
+
+    In other words, it is the orbit of the function [f] out of
+    the element [x].
+
+    @since 4.14 *)
+
+(** {1 Transforming sequences} *)
+
+(** The functions in this section are lazy: that is, they return sequences
+    whose elements are computed only when demanded. *)
+
 val map : ('a -> 'b) -> 'a t -> 'b t
-(** [map f seq] returns a new sequence whose elements are the elements of
-    [seq], transformed by [f].
-    This transformation is lazy, it only applies when the result is traversed.
+(** [map f xs] is the image of the sequence [xs] through the
+    transformation [f].
+
+    If [xs] is the sequence [x0; x1; ...] then
+    [map f xs] is the sequence [f x0; f x1; ...]. *)
 
-    If [seq = [1;2;3]], then [map f seq = [f 1; f 2; f 3]]. *)
+val mapi : (int -> 'a -> 'b) -> 'a t -> 'b t
+(** [mapi] is analogous to [map], but applies the function [f] to
+    an index and an element.
+
+    [mapi f xs] is equivalent to [map2 f (ints 0) xs].
+
+    @since 4.14 *)
 
 val filter : ('a -> bool) -> 'a t -> 'a t
-(** Remove from the sequence the elements that do not satisfy the
-    given predicate.
-    This transformation is lazy, it only applies when the result is
-    traversed. *)
+(** [filter p xs] is the sequence of the elements [x] of [xs]
+    that satisfy [p x].
+
+    In other words, [filter p xs] is the sequence [xs],
+    deprived of the elements [x] such that [p x] is false. *)
 
 val filter_map : ('a -> 'b option) -> 'a t -> 'b t
-(** Apply the function to every element; if [f x = None] then [x] is dropped;
-    if [f x = Some y] then [y] is returned.
-    This transformation is lazy, it only applies when the result is
-    traversed. *)
+(** [filter_map f xs] is the sequence of the elements [y] such that
+    [f x = Some y], where [x] ranges over [xs].
+
+    [filter_map f xs] is equivalent to
+    [map Option.get (filter Option.is_some (map f xs))]. *)
```

Comment:
`filteri` and `filter_mapi` may be valuable too. `List` already contains the former.